File: oriedita/src/main/java/com/example/oriedita/editor/handler/BaseMouseHandlerLineTransform.java
package com.example.oriedita.editor.handler;

import org.tinylog.Logger;
import oriedita.editor.databinding.AngleSystemModel;
import oriedita.editor.databinding.CanvasModel;
import oriedita.editor.drawing.tools.Camera;
import oriedita.editor.drawing.tools.DrawingUtil;
import oriedita.editor.save.Save;
import oriedita.editor.save.SaveProvider;
import origami.crease_pattern.FoldLineSet;
import origami.crease_pattern.elements.LineSegment;
import origami.crease_pattern.elements.Point;

import java.awt.Color;
import java.awt.Graphics2D;
import java.awt.image.BufferedImage;

public abstract class BaseMouseHandlerLineTransform extends BaseMouseHandlerLineSelect {

    protected CanvasModel canvasModel;
    protected FoldLineSet lines;
    protected Point delta;

    private BufferedImage image;
    private boolean cacheTooBig;
    private boolean needsRerender = false;
    protected boolean active = false;
    private double lastZoomX;
    private double lastZoomY;
    private double lastAngle;
    private Point bottomLeft, topRight;

    protected BaseMouseHandlerLineTransform(CanvasModel canvasModel, AngleSystemModel angleSystemModel) {
        super(angleSystemModel);
        this.canvasModel = canvasModel;
    }

    @Override
    public void mousePressed(Point p0) {
        super.mousePressed(p0);

        delta = new Point(0, 0);
        FoldLineSet ori_s_temp = new FoldLineSet();    //セレクトされた折線だけ取り出すために使う
        Save save = SaveProvider.createInstance();
        d.getFoldLineSet().getMemoSelectOption(save, 2);
        ori_s_temp.setSave(save);
        lines = ori_s_temp;
        active = true;
        needsRerender = true;
        bottomLeft = null;
        topRight = null;
    }

    @Override
    public void mouseDragged(Point p0) {
        super.mouseDragged(p0);
        delta = new Point(
                -selectionLine.determineBX() + selectionLine.determineAX(),
                -selectionLine.determineBY() + selectionLine.determineAY()
        );
    }

    @Override
    public void mouseReleased(Point p0) {
        //  <-------20180919この行はセレクトした線の端点を選ぶと、移動とかコピー等をさせると判断するが、その操作が終わったときに必要だから追加した。

        delta = new Point(
                -selectionLine.determineBX() + selectionLine.determineAX(),
                -selectionLine.determineBY() + selectionLine.determineAY()
        );
        d.getLineStep().clear();
        active = false;
        image = null;
        cacheTooBig = false;
    }

    @Override
    public void drawPreview(Graphics2D g2, Camera camera, DrawingSettings settings) {
        super.drawPreview(g2, camera, settings);
        if (!active) {
            return;
        }
        if (lines.getTotal() < 1000) { // no need to cache with so few lines
            drawDirect(g2, camera, settings);
            return;
        }
        if (determineCameraChanged(camera)) {
            cacheTooBig = false;
            needsRerender = true;
        }
        lastZoomX = camera.getCameraZoomX();
        lastZoomY = camera.getCameraZoomY();
        lastAngle = camera.getCameraAngle();

        if (needsRerender && lines != null && !cacheTooBig) {
            initCacheImage(g2, camera, settings);
            if (image != null) { // image won't be created if it would be too big
                rerender(camera, settings);
                needsRerender = false;
            }
        }

        if (image != null) {
            Point origin = camera.object2TV(new Point(0, 0));
            Point deltaTransformed = camera.object2TV(delta);
            Logger.info(delta);
            g2.drawImage(image,
                    (int) (bottomLeft.getX() + deltaTransformed.getX() - origin.getX()),
                    (int) (bottomLeft.getY() + deltaTransformed.getY() - origin.getY()),
                    image.getWidth(), image.getHeight(), null);
        }
        if (image == null && lines != null) {
            drawDirect(g2, camera, settings);
        }
    }

    private void initCacheImage(Graphics2D g2, Camera camera, DrawingSettings settings) {
        if (bottomLeft == null) {
            double minX = lines.getMinX();
            double maxX = lines.getMaxX();
            double minY = lines.getMinY();
            double maxY = lines.getMaxY();

            bottomLeft = camera.object2TV(new Point(minX, minY));
            topRight = camera.object2TV(new Point(maxX, maxY));
        }

        int width = (int) (topRight.getX() - bottomLeft.getX()) + 3;
        int height = (int) (topRight.getY() - bottomLeft.getY()) + 3;
        image = null;
        if (width * height < settings.getWidth() * settings.getHeight() * 1.5) {
            image = g2.getDeviceConfiguration().createCompatibleImage(width, height, BufferedImage.BITMASK);
            bottomLeft = bottomLeft.move(new Point(-1, -1));
        }
    }

    protected void drawDirect(Graphics2D g2, Camera camera, DrawingSettings settings) {
        Point origin = camera.object2TV(new Point(0, 0));
        Point deltaTransformed = camera.object2TV(delta);
        int minx = (int) -(deltaTransformed.getX() - origin.getX());
        int miny = (int) -(deltaTransformed.getY() - origin.getY());
        int maxx = minx + settings.getWidth();
        int maxy = miny + settings.getHeight();

        // do cohen-sutherland clipping
        for (var s : lines.getLineSegmentsIterable()) {
            Point a = camera.object2TV(s.getA());
            Point b = camera.object2TV(s.getB());
            int regionA = DrawingUtil.cohenSutherlandRegion(minx, miny, maxx, maxy, a);
            int regionB = DrawingUtil.cohenSutherlandRegion(minx, miny, maxx, maxy, b);
            if ((regionA & regionB) == DrawingUtil.CENTER) {
                Point pa = s.getA().move(delta);
                Point pb = s.getB().move(delta);
                LineSegment s2 = new LineSegment(pa, pb);
                DrawingUtil.drawCpLine(g2, s2, camera, settings.getLineStyle(), settings.getLineWidth(),
                        d.getPointSize(), settings.getWidth(), settings.getHeight(), settings.useRoundedEnds());
            }
        }
    }

    protected void rerender(Camera camera, DrawingSettings settings) {
        Point zero = camera.TV2object(new Point(0, 0));
        Point boObject = camera.TV2object(bottomLeft);
        FoldLineSet ori_s_temp = new FoldLineSet();
        ori_s_temp.set(lines);
        ori_s_temp.move(zero.getX() - boObject.getX(), zero.getY() - boObject.getY());
        Graphics2D g = image.createGraphics();
        g.setBackground(new Color(0f, 0, 0, 0));
        for (var ls : ori_s_temp.getLineSegmentsIterable()) {
            DrawingUtil.drawCpLine(g, ls, camera, settings.getLineStyle(),
                    settings.getLineWidth(), d.getPointSize(), image.getWidth(), image.getHeight(), settings.useRoundedEnds());
        }
    }

    private boolean determineCameraChanged(Camera camera) {
        return camera.getCameraZoomX() != lastZoomX || camera.getCameraZoomY() != lastZoomY || camera.getCameraAngle() != lastAngle;
    }

    @Override
    public void reset() {
        super.reset();
        image = null;
        needsRerender = false;
        bottomLeft = null;
        cacheTooBig = false;
        active = false;
        lastAngle = 100000;
        lastZoomX = 0;
        lastZoomY = 0;
    }
}

File: oriedita-ui/src/main/java/com/example/oriedita_ui/handler/BaseMouseHandlerLineTransform.kt
package com.example.oriedita_ui.handler

import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.input.pointer.PointerInputChange
import com.example.oriedita_core.origami.crease_pattern.elements.LineSegment
import com.example.oriedita_core.origami.crease_pattern.elements.Point
import com.example.oriedita_core.origami.crease_pattern.FoldLineSet

/**
 * Базовый класс для обработчиков мыши с трансформацией линий
 * Позволяет перемещать и трансформировать выбранные линии
 * Адаптированная версия для Android
 */
abstract class BaseMouseHandlerLineTransform : BaseMouseHandlerLineSelect() {
    
    protected var delta = Point(0.0, 0.0)
    protected var lines: FoldLineSet? = null
    protected var active = false
    
    private var needsRerender = false
    private var lastZoomX = 0.0
    private var lastZoomY = 0.0
    private var lastAngle = 0.0
    private var bottomLeft: Point? = null
    private var topRight: Point? = null
    
    override fun onPress(offset: Offset, change: PointerInputChange): Boolean {
        super.onPress(offset, change)
        
        delta = Point(0.0, 0.0)
        
        // Получить выбранные линии
        val selectedLines = getSelectedLines()
        lines = selectedLines
        active = true
        needsRerender = true
        bottomLeft = null
        topRight = null
        
        println("Начало трансформации линий: ${selectedLines.getTotal()} линий")
        return true
    }
    
    override fun onDrag(offset: Offset, change: PointerInputChange): Boolean {
        super.onDrag(offset, change)
        
        if (selectionLine != null) {
            val selectionLine = selectionLine!!
            delta = Point(
                -selectionLine.determineBX() + selectionLine.determineAX(),
                -selectionLine.determineBY() + selectionLine.determineAY()
            )
        }
        
        return true
    }
    
    override fun onRelease(offset: Offset, change: PointerInputChange): Boolean {
        if (selectionLine != null) {
            val selectionLine = selectionLine!!
            delta = Point(
                -selectionLine.determineBX() + selectionLine.determineAX(),
                -selectionLine.determineBY() + selectionLine.determineAY()
            )
        }
        
        clearLineStep()
        active = false
        
        println("Завершение трансформации линий: дельта (${delta.x}, ${delta.y})")
        return true
    }
    
    /**
     * Получить выбранные линии для трансформации
     */
    protected fun getSelectedLines(): FoldLineSet {
        val selectedLines = FoldLineSet()
        // Здесь должна быть логика получения выбранных линий
        // Пока возвращаем пустой набор
        return selectedLines
    }
    
    /**
     * Применить трансформацию к линиям
     */
    protected fun applyTransformation() {
        lines?.let { foldLineSet ->
            for (line in foldLineSet.getLineSegmentsIterable()) {
                val newA = line.getA().move(delta)
                val newB = line.getB().move(delta)
                val transformedLine = LineSegment(newA, newB, line.getColor())
                // Здесь должна быть логика применения трансформации
            }
        }
    }
    
    /**
     * Проверить, изменилась ли камера
     */
    private fun determineCameraChanged(zoomX: Double, zoomY: Double, angle: Double): Boolean {
        return zoomX != lastZoomX || zoomY != lastZoomY || angle != lastAngle
    }
    
    /**
     * Обновить кэш рендеринга
     */
    protected fun updateRenderCache(zoomX: Double, zoomY: Double, angle: Double) {
        if (determineCameraChanged(zoomX, zoomY, angle)) {
            needsRerender = true
        }
        
        lastZoomX = zoomX
        lastZoomY = zoomY
        lastAngle = angle
    }
    
    /**
     * Получить ограничивающий прямоугольник линий
     */
    protected fun getBoundingBox(): BoundingBox? {
        lines?.let { foldLineSet ->
            if (foldLineSet.getTotal() > 0) {
                try {
                    val minX = foldLineSet.getMinX()
                    val maxX = foldLineSet.getMaxX()
                    val minY = foldLineSet.getMinY()
                    val maxY = foldLineSet.getMaxY()
                    
                    // Проверяем, что координаты валидны
                    if (minX.isFinite() && maxX.isFinite() && minY.isFinite() && maxY.isFinite() &&
                        minX <= maxX && minY <= maxY) {
                        
                        return BoundingBox(
                            topLeft = Point(minX, minY),
                            bottomRight = Point(maxX, maxY)
                        )
                    } else {
                        // Если координаты невалидны, вычисляем вручную
                        return calculateBoundingBoxManually(foldLineSet)
                    }
                } catch (e: Exception) {
                    // Если методы не работают, вычисляем вручную
                    return calculateBoundingBoxManually(foldLineSet)
                }
            }
        }
        return null
    }
    
    /**
     * Вычисление ограничивающего прямоугольника вручную
     */
    private fun calculateBoundingBoxManually(foldLineSet: FoldLineSet): BoundingBox? {
        if (foldLineSet.getTotal() <= 0) return null
        
        var minX = Double.POSITIVE_INFINITY
        var maxX = Double.NEGATIVE_INFINITY
        var minY = Double.POSITIVE_INFINITY
        var maxY = Double.NEGATIVE_INFINITY
        
        try {
            for (i in 1..foldLineSet.getTotal()) {
                val line = foldLineSet.get(i)
                val ax = line.determineAX()
                val ay = line.determineAY()
                val bx = line.determineBX()
                val by = line.determineBY()
                
                minX = minOf(minX, ax, bx)
                maxX = maxOf(maxX, ax, bx)
                minY = minOf(minY, ay, by)
                maxY = maxOf(maxY, ay, by)
            }
            
            // Проверяем, что координаты валидны
            if (minX.isFinite() && maxX.isFinite() && minY.isFinite() && maxY.isFinite() &&
                minX <= maxX && minY <= maxY) {
                
                return BoundingBox(
                    topLeft = Point(minX, minY),
                    bottomRight = Point(maxX, maxY)
                )
            }
        } catch (e: Exception) {
            println("Ошибка при вычислении ограничивающего прямоугольника: ${e.message}")
        }
        
        return null
    }
    

    
    fun getLinesCount(): Int = lines?.getTotal() ?: 0
    
    override fun reset() {
        super.reset()
        delta = Point(0.0, 0.0)
        lines = null
        active = false
        needsRerender = false
        bottomLeft = null
        topRight = null
        lastZoomX = 0.0
        lastZoomY = 0.0
        lastAngle = 0.0
        
        println("Обработчик трансформации линий сброшен")
    }
    
    fun getLineTransformDescription(): String {
        return when {
            !active -> "Готов к трансформации"
            lines == null -> "Трансформация..."
            else -> "Трансформация ${lines!!.getTotal()} линий: дельта (${String.format("%.2f", delta.x)}, ${String.format("%.2f", delta.y)})"
        }
    }
    
    fun getLineTransformInfo(): String {
        return buildString {
            append("Статус: ${if (active) "активна" else "неактивна"}\n")
            append("Линий: ${getLinesCount()}\n")
            append("Дельта: (${String.format("%.4f", delta.x)}, ${String.format("%.4f", delta.y)})\n")
            append("Нужен перерендер: ${if (needsRerender) "да" else "нет"}")
        }
    }
}

data class BoundingBox(
    val topLeft: Point,
    val bottomRight: Point
) {
    val width: Double get() = bottomRight.x - topLeft.x
    val height: Double get() = bottomRight.y - topLeft.y
    val center: Point get() = Point((topLeft.x + bottomRight.x) / 2, (topLeft.y + bottomRight.y) / 2)
    
    fun contains(point: Point): Boolean {
        return point.x >= topLeft.x && point.x <= bottomRight.x &&
               point.y >= topLeft.y && point.y <= bottomRight.y
    }
    
    fun intersects(other: BoundingBox): Boolean {
        return !(other.bottomRight.x < topLeft.x || other.topLeft.x > bottomRight.x ||
                other.bottomRight.y < topLeft.y || other.topLeft.y > bottomRight.y)
    }
} 